language/move-core/types/src/value.rs

```diff
@@ -144,6 +192,38 @@ impl<'d, 'a> serde::de::Visitor<'d> for VectorElementVisitor<'a> {
     }
 }
 
+struct DecoratedStructFieldVisitor<'a>(&'a [(Identifier, MoveTypeLayout)]);
+
+impl<'d, 'a> serde::de::Visitor<'d> for DecoratedStructFieldVisitor<'a> {
+    type Value = BTreeMap<Identifier, MoveValue>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter.write_str("Struct")
+    }
+
+    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::MapAccess<'d>,
+    {
+        let mut val: BTreeMap<Identifier, MoveValue> = BTreeMap::new();
+        // TODO: is this iteration order deterministic? should use a vec of tuples if not
+        for (i, (f, layout)) in self.0.iter().enumerate() {
+            match map.next_entry_seed(PhantomData::default(), layout)? {
+                Some((other_f, elem)) => {
+                    if f != &other_f {
+                        return Err(A::Error::custom(
+                            "Fields ordered differently in layout vs struct",
+                        ));
+                    }
+                    val.insert(other_f, elem);
+                }
+                None => return Err(A::Error::invalid_length(i, &self)),
+            }
+        }
+        Ok(val)
```

Comment:
Honestly, this doesn't look quite right to me. Correct me if I'm wrong, but I don't think we are requiring fields to be alphabetically sorted and using a BtreeMap indexed by field names will break the original ordering. A `Vec<(Identifier, MoveValue)>` seems to be the better choice here...

Reply:
Good point. I'll go with the vec here.